### ISSUE TITLE
Prompt_ids vs. decoder_input_ids in Whisper

### ISSUE THREAD

**Issue by vymao** [huggingface/transformers#28228]
### Feature request

I am trying to understand the different between adding prior text to `prompt_ids` vs. `decoder_input_ids` when [generating text via Whisper](https://huggingface.co/docs/transformers/v4.36.1/en/model_doc/whisper#transformers.WhisperForConditionalGeneration). The documentation is not very clear on how these differ implementation-wise; AFAIK, it seems like using `prompt_ids` will lead to `forced_input_ids` being modified [here](https://github.com/huggingface/transformers/blob/v4.36.1/src/transformers/models/whisper/modeling_whisper.py#L2211-L2215). But I'm not sure how exactly using `decoder_input_ids` differs from this.

### Motivation

To add context to the whisper transcription. For example, if the model previously transcribed `I have a` in a streaming fashion, I would like to add this as "context" into the model to help it predict the next word. I believe the actual OpenAI Whisper implementation has a feature called "prefix" that does this.

### Your contribution

Will try.

**Comment by bL34cHig0** [comment]
I'd like to undertake this issue please

**Comment by vymao** [comment]
Thanks. I'm still slightly confused: when you say `prompt_ids` are used to provide initial context, isn't that still on the decoder side before the actual generated text? How is this different from using `decoder_input_ids`?

**Comment by LysandreJik** [comment]
Maybe for @sanchit-gandhi or @ylacombe

**Comment by ylacombe** [comment]
Hey @vymao, I'm not a Whisper expert yet but as I understand and as the [documentation](https://huggingface.co/docs/transformers/v4.36.1/en/model_doc/whisper#transformers.WhisperForConditionalGeneration.generate.prompt_ids) suggests, `prompt_ids` are created by using the tokenizer's or the processor's `get_prompt_ids`.

https://github.com/huggingface/transformers/blob/3cefac1d974db5e2825a0cb2b842883a628be7a0/src/transformers/models/whisper/tokenization_whisper.py#L827-L836

As you can see from the code, get_prompt_ids handles the input text so you don't have to worry about special tokens that need to be inserted to tell the model that this text is context and not the start of the transcription.

Then the code processes the `prompt_ids` in place of the `decoder_input_ids`.

In other words, you can use  `prompt_ids` obtained from `get_prompt_ids` if you want to pass a context to Whisper.  `decoder_input_ids` is much more flexible: you could reproduce `prompt_ids` obtained from `get_prompt_ids` or use it to have a more advanced use of Whisper.

I hope that it helps!

cc @sanchit-gandhi or @ArthurZucker if you want to correct me or give a more advanced explanation
